Allow clients to pass in explicit start and end timestamps to spans

jvans1 opened the pull request:
The [open tracing spec](https://github.com/opentracing/specification/blob/master/specification.md) allows for the clients to pass an optional start and end timestamp to spans. This change enables that option for clients while providing the default of the current wall time as defined in the spec.

adriancole commented:
opentracing is dead, but anyway this isn't an opentracing project. please don't cite OT as rationale for anything.

What are you aiming to do with this? Sometimes people are trying to record spans after the fact, and there are better ways to do that.

jvans1 commented:
Sorry I didn't realize this wasn't an OT implementation, a lot of the vernacular is the same but I guess they differ in some ways. What is the best way to record a span after the fact? I've dug into most of the code and had a hard time finding a clean way to do that.

adriancole commented:
I meant to use the sender directly, like this for the out-of-band spans

https://github.com/openzipkin/zipkin-ruby/blob/master/lib/zipkin-tracer/zipkin_http_sender.rb

I'm not sure how much lifecycle is implicit in the sender api. In other projects they just take object literals, but this one seems to invoke lifecycle unless I'm mistaken @ykitamura-mdsol 
https://github.com/openzipkin/zipkin-ruby/blob/9cc7ba6b42955e7f7fff9528c40d90e8071508a1/lib/zipkin-tracer/zipkin_sender_base.rb

anyway it would seem not a terribly hard refactor to allow senders to accept complete span objects, ideally plain ones like structs

jcarres-mdsol commented:
@jvans1 still interested in this? I'd merge, add a line to the changelog and update the version

jvans1 commented:
@jcarres-mdsol yes! I've added a line to the changelog and bumped the version. I noticed the travis build was failing and it looks related to some [openjdk issue](https://travis-ci.community/t/error-installing-oraclejdk8-expected-feature-release-number-in-range-of-9-to-14-but-got-8/3766/6) on travis. I added their recommended fix of specifying `dist: trusty` in the travis.yml and it seems to have fixed it

jcarres-mdsol commented:
Looks great. Thanks!

jcarres-mdsol commented:
Thanks for fixing the failing build !
